File: sdk/ruby/lib/cordum/cap/worker.rb
# frozen_string_literal: true

module Cordum
  module Cap
    class Worker
      # @param nc [NATS::IO::Client] connected NATS client
      # @param subject [String] NATS subject to subscribe to
      # @param sender_id [String] this worker's sender ID
      # @param handler [Proc] callable (ctx, request) -> JobResult
      # @param private_key [OpenSSL::PKey::EC, nil] optional signing key
      # @param public_keys [Hash{String => OpenSSL::PKey::EC}] sender_id => key map
      # @param metrics [MetricsHook] metrics callback object
      # @param queue_group [String] NATS queue group for load balancing
      def initialize(nc:, subject:, sender_id:, handler:,
                     private_key: nil, public_keys: {}, metrics: NoopMetrics.new,
                     queue_group: 'workers')
        @nc = nc
        @subject = subject
        @sender_id = sender_id
        @handler = handler
        @private_key = private_key
        @public_keys = public_keys
        @metrics = metrics
        @queue_group = queue_group
        @middlewares = []
        @running = false
      end

      # Adds middleware(s) to the processing chain.
      def use(*middlewares)
        @middlewares.concat(middlewares)
        self
      end

      # Starts the worker subscription. Blocks until stop is called.
      def start
        wrapped = MiddlewareChain.apply(@handler, @middlewares)
        @running = true

        @sid = @nc.subscribe(@subject, queue: @queue_group) do |msg|
          process_message(msg.data, wrapped)
        end

        while @running
          sleep 0.1
        end
      end

      # Stops the worker.
      def stop
        @running = false
        @nc.unsubscribe(@sid) if @sid
      end

      private

      def process_message(data, handler)
        pkt = Cordum::Agent::V1::BusPacket.decode(data)

        # Verify signature if public keys are configured
        if !@public_keys.empty? && !pkt.sender_id.empty?
          pub_key = @public_keys[pkt.sender_id]
          if pub_key && !Signing.verify_packet_signature(pkt, pub_key)
            @metrics.on_error('signature', "invalid signature from #{pkt.sender_id}")
            return
          end
        end

        return unless pkt.payload == :job_request

        req = pkt.job_request
        @metrics.on_job_received(req.job_id, req.topic)

        ctx = Context.new(trace_id: pkt.trace_id, sender_id: pkt.sender_id)
        start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)

        begin
          result = handler.call(ctx, req)
          elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond) - start_time
          @metrics.on_job_completed(req.job_id, elapsed, result.status.to_s)
        rescue StandardError => e
          elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond) - start_time
          @metrics.on_job_failed(req.job_id, e.message)
          result = Cordum::Agent::V1::JobResult.new(
            job_id: req.job_id,
            worker_id: @sender_id,
            status: :JOB_STATUS_FAILED,
            error_code: e.is_a?(CapError) ? e.code.to_s : 'UNKNOWN',
            error_message: e.message,
            execution_ms: elapsed
          )
        end

        reply_pkt = Cordum::Agent::V1::BusPacket.new(
          trace_id: pkt.trace_id,
          sender_id: @sender_id,
          protocol_version: DEFAULT_PROTOCOL_VERSION,
          created_at: Google::Protobuf::Timestamp.new(seconds: Time.now.to_i, nanos: 0),
          job_result: result
        )

        Signing.sign_packet(reply_pkt, @private_key) if @private_key

        reply_data = Codec.marshal_deterministic(reply_pkt)
        @nc.publish(SUBJECT_RESULT, reply_data)
      rescue StandardError => e
        @metrics.on_error('process', e.message)
      end
    end
  end
end

File: sdk/dotnet/src/Cordum.Cap/Worker.cs
using System.Diagnostics;
using System.Security.Cryptography;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using NATS.Client.Core;
using Cordum.Agent.V1;

namespace Cordum.Cap;

/// <summary>
/// CAP worker that processes jobs from a NATS subject.
/// </summary>
public sealed class Worker
{
    private readonly NatsConnection _nc;
    private readonly string _subject;
    private readonly string _senderId;
    private readonly HandlerDelegate _baseHandler;
    private readonly ECDsa? _privateKey;
    private readonly Dictionary<string, ECDsa> _publicKeys;
    private readonly IMetricsHook _metrics;
    private readonly List<MiddlewareDelegate> _middlewares = new();

    public Worker(
        NatsConnection nc,
        string subject,
        string senderId,
        HandlerDelegate handler,
        ECDsa? privateKey = null,
        Dictionary<string, ECDsa>? publicKeys = null,
        IMetricsHook? metrics = null)
    {
        _nc = nc;
        _subject = subject;
        _senderId = senderId;
        _baseHandler = handler;
        _privateKey = privateKey;
        _publicKeys = publicKeys ?? new Dictionary<string, ECDsa>();
        _metrics = metrics ?? new NoopMetrics();
    }

    /// <summary>
    /// Add middleware(s) to the processing chain.
    /// </summary>
    public void Use(params MiddlewareDelegate[] mws)
    {
        _middlewares.AddRange(mws);
    }

    /// <summary>
    /// Start processing messages. Blocks until cancellation is requested.
    /// </summary>
    public async Task StartAsync(CancellationToken ct = default)
    {
        var handler = MiddlewareChain.ApplyMiddleware(_baseHandler, _middlewares);

        await foreach (var msg in _nc.SubscribeAsync<byte[]>(_subject, queueGroup: _subject, cancellationToken: ct))
        {
            if (msg.Data == null) continue;

            BusPacket packet;
            try
            {
                packet = BusPacket.Parser.ParseFrom(msg.Data);
            }
            catch (InvalidProtocolBufferException e)
            {
                _metrics.OnError("parse", $"failed: {e.Message}");
                continue;
            }

            if (_publicKeys.Count > 0)
            {
                if (!_publicKeys.TryGetValue(packet.SenderId, out var pubKey))
                    continue;
                if (packet.Signature.IsEmpty || !Signing.VerifyPacketSignature(packet, pubKey))
                    continue;
            }

            if (packet.PayloadCase != BusPacket.PayloadOneofCase.JobRequest)
                continue;

            var req = packet.JobRequest;
            _metrics.OnJobReceived(req.JobId, req.Topic);

            var ctx = new Context
            {
                TraceId = packet.TraceId,
                SenderId = packet.SenderId,
                Packet = packet,
            };

            var sw = Stopwatch.StartNew();
            JobResult result;
            try
            {
                result = handler(ctx, req);
                if (string.IsNullOrEmpty(result.JobId)) result.JobId = req.JobId;
                if (string.IsNullOrEmpty(result.WorkerId)) result.WorkerId = _senderId;

                sw.Stop();
                if (result.Status == JobStatus.Failed)
                {
                    _metrics.OnJobFailed(result.JobId, result.ErrorMessage);
                }
                else
                {
                    _metrics.OnJobCompleted(result.JobId, sw.ElapsedMilliseconds, "SUCCEEDED");
                }
            }
            catch (Exception ex)
            {
                sw.Stop();
                _metrics.OnJobFailed(req.JobId, ex.Message);
                result = new JobResult
                {
                    JobId = req.JobId,
                    WorkerId = _senderId,
                    Status = JobStatus.Failed,
                    ErrorMessage = ex.Message,
                };
            }

            var outPacket = new BusPacket
            {
                TraceId = packet.TraceId,
                SenderId = _senderId,
                ProtocolVersion = Subjects.DefaultProtocolVersion,
                CreatedAt = Timestamp.FromDateTimeOffset(DateTimeOffset.UtcNow),
                JobResult = result,
            };

            if (_privateKey != null)
            {
                Signing.SignPacket(outPacket, _privateKey);
            }

            var data = Codec.MarshalDeterministic(outPacket);
            await _nc.PublishAsync(Subjects.Result, data, cancellationToken: ct);
        }
    }
}
